modules/setting/storage.go

```diff
@@ -4,87 +4,182 @@
 	// Global Defaults
-	sec.Key("MINIO_ENDPOINT").MustString("localhost:9000")
-	sec.Key("MINIO_ACCESS_KEY_ID").MustString("")
-	sec.Key("MINIO_SECRET_ACCESS_KEY").MustString("")
-	sec.Key("MINIO_BUCKET").MustString("gitea")
-	sec.Key("MINIO_LOCATION").MustString("us-east-1")
-	sec.Key("MINIO_USE_SSL").MustBool(false)
-	sec.Key("MINIO_INSECURE_SKIP_VERIFY").MustBool(false)
-	sec.Key("MINIO_CHECKSUM_ALGORITHM").MustString("default")
+	storageSec.Key("STORAGE_TYPE").MustString("local")
+	storageSec.Key("MINIO_ENDPOINT").MustString("localhost:9000")
+	storageSec.Key("MINIO_ACCESS_KEY_ID").MustString("")
+	storageSec.Key("MINIO_SECRET_ACCESS_KEY").MustString("")
+	storageSec.Key("MINIO_BUCKET").MustString("gitea")
+	storageSec.Key("MINIO_LOCATION").MustString("us-east-1")
+	storageSec.Key("MINIO_USE_SSL").MustBool(false)
+	storageSec.Key("MINIO_INSECURE_SKIP_VERIFY").MustBool(false)
+	storageSec.Key("MINIO_CHECKSUM_ALGORITHM").MustString("default")
+	return storageSec
+}
 
-	if targetSec == nil {
-		targetSec, _ = rootCfg.NewSection(name)
+func getStorage(rootCfg ConfigProvider, name, typ string, sec ConfigSection) (*Storage, error) {
+	if name == "" {
+		return nil, errors.New("no name for storage")
 	}
 
-	var storage Storage
-	storage.Section = targetSec
-	storage.Type = typ
-
-	overrides := make([]ConfigSection, 0, 3)
-	nameSec, err := rootCfg.GetSection(sectionName + "." + name)
-	if err == nil {
-		overrides = append(overrides, nameSec)
+	var targetSec ConfigSection
+	if typ != "" {
+		var err error
+		targetSec, err = rootCfg.GetSection(storageSectionName + "." + typ)
+		if err != nil {
+			if !IsValidStorageType(StorageType(typ)) {
+				return nil, fmt.Errorf("get section via storage type %q failed: %v", typ, err)
+			}
+		}
+		if targetSec != nil {
+			targetType := targetSec.Key("STORAGE_TYPE").String()
+			if targetType == "" {
+				if !IsValidStorageType(StorageType(typ)) {
+					return nil, fmt.Errorf("unknow storage type %q", typ)
+				}
+				targetSec.Key("STORAGE_TYPE").SetValue(typ)
+			} else if !IsValidStorageType(StorageType(targetType)) {
+				return nil, fmt.Errorf("unknow storage type %q for section storage.%v", targetType, typ)
+			}
+		}
 	}
 
-	typeSec, err := rootCfg.GetSection(sectionName + "." + typ)
-	if err == nil {
-		overrides = append(overrides, typeSec)
-		nextType := typeSec.Key("STORAGE_TYPE").String()
-		if len(nextType) > 0 {
-			storage.Type = nextType // Support custom STORAGE_TYPE
+	storageNameSec, _ := rootCfg.GetSection(storageSectionName + "." + name)
+
+	if targetSec == nil {
+		targetSec = sec
+	}
+	if targetSec == nil {
+		targetSec = storageNameSec
+	}
+	if targetSec == nil {
+		targetSec = getDefaultStorageSection(rootCfg)
+	} else {
+		targetType := targetSec.Key("STORAGE_TYPE").String()
+		switch {
+		case targetType == "":
+			if targetSec.Key("PATH").String() == "" {
+				targetSec = getDefaultStorageSection(rootCfg)
+			} else {
+				targetSec.Key("STORAGE_TYPE").SetValue("local")
```

Comment:
```
[storage]
STORAGE_TYPE = minio
[storage.sub]
; no STORAGE_TYPE
```

This `SetValue` will overwrite STORAGE_TYPE in `[storage]` section.

Reply:
Because targetSec's STORAGE_TYPE is empty, so we can do it safely. See line 125-127

Reply:
```
[storage]
STORAGE_TYPE = minio
[storage.sub1]
; no STORAGE_TYPE
[storage.sub2]
; no STORAGE_TYPE
```

Then if you overwrite the `STORAGE_TYPE` for `[storage.sub1]`, it also affects `[storage.sub2]`, is it intentional?

Reply:
No. It will not affect sub2. This logic will help to fill the storage type for this situation when `PATH` is not empty, then we will set `STORAGE_TYPE` to `local`.

```ini
[storage.lfs]
PATH =
```

Reply:
Suppose: 

```
[storage]
STORAGE_TYPE = minio
[storage.sub1]
; no STORAGE_TYPE
[storage.sub2]
; no STORAGE_TYPE
```

Then `targetSec.Key("STORAGE_TYPE")` points to `[storage].STORAGE_TYPE`, then you set it to "local"

Now the INI becomes:

```
[storage]
STORAGE_TYPE = local
[storage.sub1]
; no STORAGE_TYPE
[storage.sub2]
; no STORAGE_TYPE
```


I still do not think it is right.

Reply:
The set only when `STORAGE_TYPE` is empty and `PATH` is not empty. In your example, STORAGE_TYPE is not empty.

Reply:
Hmm, maybe my example is not appropriate. But I can see using `Key` + `SetValue` in sub section is fragile and maybe even dangerous. Hopefully no user would be affected by it. Fortunately, it could be bypassed by writing everything explicitly.